thetis/diagnostics.py

```diff
@@ -140,10 +154,36 @@ def __init__(self, field_2d, hessian_2d, gradient_2d=None, **kwargs):
         prob = LinearVariationalProblem(a, L, sol)
         kwargs.setdefault('solver_parameters', sp)
         self.solver = LinearVariationalSolver(prob, **kwargs)
+        self._isfrozen = True
 
     @PETSc.Log.EventDecorator("thetis.HessianRecoverer2D.solve")
     def solve(self):
         self.solver.solve()
         self.hessian_2d.assign(self._hessian)
         if self.gradient_2d is not None:
             self.gradient_2d.assign(self._gradient)
+
+
+class DynamicPressureCalculator(FrozenHasTraits):
+    r"""
+    Class for calculating dynamic pressure (i.e. kinetic energy).
+    """
+    density = FiredrakeScalarExpression(
+        Constant(1030.0), help='Fluid density').tag(config=True)
+
+    @PETSc.Log.EventDecorator("thetis.DynamicPressureCalculator.__init__")
+    def __init__(self, uv, dp, density=Constant(1030.0)):
```

Comment:
I guess by default we should use `physical_constants['rho0']` as the density value — it is the value the 2D solver uses internally. In the 3D baroclinic case the user would need to plug in the dynamical density expression `rho_0 + solver_obj.fields.density_3d` but I guess we don't need to specify it in this routine.